src/core/test_helpers/so_migrations.ts

```diff
@@ -0,0 +1,175 @@
+ *       expect.objectContaining({ type: 'my-type', id: 'my-id', attributes: { ... }, references: [] })
+ *     ]);
+ *   });
+ * });
+ * ```
+ */
+export const createTestHarness = () => {
+  let started = false;
+  let stopped = false;
+  let esServer: kbnTestServer.TestElasticsearchUtils;
+  const { startES } = kbnTestServer.createTestServers({ adjustTimeout: jest.setTimeout });
+  const root = kbnTestServer.createRootWithCorePlugins({}, { oss: false });
+
+  /**
+   * Imports an array of objects into Kibana and applies migrations before persisting to Elasticsearch. Will overwrite
+   * any existing objects with the same id.
+   * @param objects
+   */
+  const importObjects = async (objects: SavedObject[]) => {
+    if (!started)
+      throw new Error(`SavedObjectTestHarness must be started before objects can be imported`);
+    if (stopped) throw new Error(`SavedObjectTestHarness cannot import objects after stopped`);
+
+    const response = await kbnTestServer
+      // Always use overwrite=true flag so we can isolate this harness to migrations
+      .getSupertest(root, 'post', '/api/saved_objects/_import?overwrite=true')
+      .set('Content-Type', 'multipart/form-data; boundary=EXAMPLE')
+      .send(
+        [
+          '--EXAMPLE',
+          'Content-Disposition: form-data; name="file"; filename="export.ndjson"',
+          'Content-Type: application/ndjson',
+          '',
+          ...objects.map((o) => JSON.stringify(o)),
+          '--EXAMPLE--',
+        ].join('\r\n')
+      )
+      .expect(200);
+
+    if (response.body.errors?.length > 0) {
+      throw new Error(
+        `Errors importing objects: ${JSON.stringify(response.body.errors, undefined, 2)}`
+      );
+    }
+  };
+
+  /**
+   * Exports objects from Kibana with all migrations applied.
+   * @param options
+   */
+  const exportObjects = async (options: ExportOptions): Promise<SavedObject[]> => {
+    if (!started)
+      throw new Error(`SavedObjectTestHarness must be started before objects can be imported`);
+    if (stopped) throw new Error(`SavedObjectTestHarness cannot import objects after stopped`);
+
+    const response = await kbnTestServer
+      .getSupertest(root, 'post', '/api/saved_objects/_export')
+      .send({
+        ...options,
+        excludeExportDetails: true,
+      })
+      .expect(200);
+
+    // Parse ndjson response
+    return response.text.split('\n').map((s: string) => JSON.parse(s));
+  };
+
+  return {
+    /**
+     * Start Kibana and Elasticsearch for migration testing. Must be called before `migrate`.
+     * In most cases, this can be called during your test's `beforeAll` hook and does not need to be called for each
+     * individual test.
+     */
+    start: async () => {
+      if (started)
+        throw new Error(`SavedObjectTestHarness already started! Cannot call start again`);
+      if (stopped)
+        throw new Error(`SavedObjectTestHarness already stopped! Cannot call start again`);
+
+      started = true;
```

Comment:
Shouldn't `started = true` expression be the last operation when we are sure everything is settled down? The same for `stopped` flag in `stop` lifecycle.